EFCore.CheckConstraints/Internal/EnumCheckConstraintConvention.cs

```diff
@@ -41,33 +42,52 @@ public virtual void ProcessModelFinalizing(IConventionModelBuilder modelBuilder,
                 var typeMapping = (RelationalTypeMapping?)property.FindTypeMapping()
                     ?? _typeMappingSource.FindMapping((IProperty)property);
                 var propertyType = Nullable.GetUnderlyingType(property.ClrType) ?? property.ClrType;
-                if (propertyType.IsEnum
-                    && typeMapping != null
-                    && !propertyType.IsDefined(typeof(FlagsAttribute), true)
-                    && property.GetColumnName(tableIdentifier) is { } columnName)
+                if (!propertyType.IsEnum
+                    || typeMapping == null
+                    || propertyType.IsDefined(typeof(FlagsAttribute), true)
+                    || property.GetColumnName(tableIdentifier) is not {} columnName)
                 {
-                    var enumValues = Enum.GetValues(propertyType);
-                    if (enumValues.Length <= 0)
-                    {
-                        continue;
-                    }
+                    continue;
+                }
+
+                var enumValues = Enum.GetValues(propertyType).Cast<object>().Select(typeMapping.GenerateSqlLiteral).ToList();
+                if (enumValues.Count == 0)
+                {
+                    continue;
+                }
 
-                    sql.Clear();
+                sql.Clear();
 
-                    sql.Append(_sqlGenerationHelper.DelimitIdentifier(columnName));
+                sql.Append(_sqlGenerationHelper.DelimitIdentifier(columnName));
+
+                // the list is a contiguous range if the following are true
+                //  first value is int x
+                //  last value is int y
+                //  y - x is one less than the number of total values
+                if (int.TryParse(enumValues[0], out var firstValue)
```

Comment:
I think it's safer/more correct to do these checks on the actual enum values as returned by Enum.GetValues, rather than reparsing the SQL literals returned by GenerateSqlLiteral back into ints.

So I'd first call Enum.GetValues, and if that returns an int array, I'd sort it and then check the first/last etc. Only if all that is fine, I'd then generate the SQL literals and render the SQL out.

Reply:
Since C# enums are backed by ints, wouldn't that always be true? The reason I do the check after the conversion is if the conversion returns the literal names of the values (e.g. for the `Value_converter` test), then they wouldn't be represented as ints in the database.